### PR TITLE
[State Restoration] Fix type issue with RestorableNumN (and its subclasses)

### PR DESCRIPTION
**Pull request by shihaohong**
## Description

This was causing problems when I was trying to migrate the gallery to use RestorableIntN because the value was cast as a `num?` instead of an `int?`. This fixes the issue. However, I'm not sure how to write a test. 

The following was my first instinct, but didn't work:
```dart
  testWidgets('RestorableN types are properly defined', (WidgetTester tester) async {
    await tester.pumpWidget(const RootRestorationScope(
      restorationId: 'root-child',
      child: _RestorableWidget(),
    ));

    expect(find.text('hello world'), findsOneWidget);
    final _RestorableWidgetState state = tester.state(find.byType(_RestorableWidget));

    state.setProperties(() {
      state.nullableIntValue.value = 24;
    });

    expect(state.nullableIntValue.value, isA<int>()); // this passes even without the fix.
  });
```
This doesn't work because it passes without the fix because `int` is a subclass of `num`, and the check considers superclasses as a match.

## Tests
-  Not sure how to, need help with this.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by shihaohong** [comment]
The way I ran into this problem was with a similar function in the gallery:
```dart
    void takesInt(int? value) {}
    takesInt(state.nullableIntValue.value);
```

Before this change, the code above results in the following Dart error:
`The argument type 'num?' can't be assigned to the parameter type 'int?'. [argument_type_not_assignable](https://dart.dev/tools/diagnostic-messages#argument_type_not_assignable)`

Maybe I could just write those functions into the test and just have them there with a comment describing why this should work?

File: packages/flutter/lib/src/widgets/restoration_properties.dart [hunk 1]
```diff
@@ -306,11 +306,11 @@ class RestorableBoolN extends _RestorablePrimitiveValueN<bool?> {
 /// See also:
 ///
 ///  * [RestorableNum] for the non-nullable version of this class.
-class RestorableNumN<T extends num?> extends _RestorablePrimitiveValueN<num?> {
+class RestorableNumN<T extends num?> extends _RestorablePrimitiveValueN<T> {
   /// Creates a [RestorableNumN].
   ///
   /// {@macro flutter.widgets.RestorableNum.constructor}
-  RestorableNumN(num? defaultValue) : super(defaultValue);
+  RestorableNumN(T defaultValue) : super(defaultValue);
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/restoration_properties.dart:313]: So this `T` should be only a num, int, or double right?  It's not possible to check that with an assertion is it?  I know Dart doesn't have union types.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

I don't know enough about how this restorable stuff works, but as a guess, could you test it by changing the restorable value to `null` now since int and double weren't nullable before this PR?

If not, I wonder what exactly was the problem with the gallery that required this change.  Maybe there's some way to set up a test that would cause the same problem.

If nothing works then I guess this is fine without a test.

*Update:* I missed your comment when I first wrote this.  Ok I see how it was failing due to the parameter type in takesInt.  How about this to avoid isA allowing a superclass?

```dart
//expect(state.nullableIntValue.value, isA<int>());
expect(state.nullableIntValue.value.runtimeType, 'int');
```

File: packages/flutter/lib/src/widgets/restoration_properties.dart (hunk 1 above)
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/restoration_properties.dart:313]: Yeah, I'm not sure if we can check it with an assertion. The way I've been enforcing the types is through the definition of T in the class, like `RestorableNumN<T extends num?>`

**Comment by shihaohong** [comment]
I think `expect(state.nullableIntValue.value.runtimeType, 'int');` doesn't work as well because I think Dart automatically casts the runtimeType into an `int` when you set a num value to an integer value (ie `value = 14;`).

In any case, I still added a tentative test with some comments explaining how the code works. It feels a little weird, but would work in preventing regressions. This is what my IDE shows when I add the test without my fix to the `master` branch, which was the error I was seeing in the gallery change I'm trying to make:
Screen Shot 2020-12-31 at 9 11 21 AM

File: packages/flutter/lib/src/widgets/restoration_properties.dart (hunk 1 above)
Comment (shihaohong) [reply packages/flutter/lib/src/widgets/restoration_properties.dart:313]: Plus, I don't think we'll have to worry about T being something other than num, int or double. According the num documentation:
 * It is a compile-time error for any type other than [int] or [double] to attempt to extend or implement num.

**Pull request merged**